### ISSUE TITLE
How to run gemm:avx2

### ISSUE THREAD

**Issue by daydayfun** [oneapi-src/oneDNN#762]
I run pytorch convolution by CPU
export DNNL_VERBOSE=1
python test_ops.py

debug info:
mkldnn_verbose,create,convlotion,gemm:blas,forward_training....

I want to run in gemm::avx2  to improve perfermance

**Comment by rsdubtso** [comment]
Not sure I completely understand the question but oneDNN provides sgemm and 8-bit gemm interfaces. 

For example, C API sgemm:
https://github.com/oneapi-src/oneDNN/blob/f5997b5e6726de82d19ae9b86b08d80aea4af82e/include/dnnl.h#L3451-L3453

**Comment by daydayfun** [comment]
Thanks,  If make MKLDNN  on ubantu 18.04, how to set ARC::AVX2 ?
I guess my cpu_ist_t of mkldnn is isa_any , and results in the bad performance of convOp

**Comment by rsdubtso** [comment]
I still feel like I does not really understand the question. oneDNN dispatches automatically to the best available implementation for the highest ISA available on the platform. This includes GEMM, it may be just not shown in the verbose output due to the way it is implemented. The first few lines of verbose output should show the detected ISA.

**Comment by vpirogov** [comment]
It seems like the question is why convolution implementation for this specific problem uses `gemm::blas` algorithm and not jit::avx2.

**Comment by vpirogov** [comment]
@daydayfun,
 
oneDNN dispatches the best available implementation for particular problem. In your particular case it seems that JIT implementation for the problem your are running is not available and oneDNN resorts to `gemm:blas` implementation. Typically this happens in "corner case" scenarios, like padding size exceeding half of the kernel size, large strides, or large kernels. If you share complete verbose line we may be able to diagnose this further.

**Comment by daydayfun** [comment]
> It seems like the question is why convolution implementation for this specific problem uses `gemm::blas` algorithm and not `jit::avx2`.

Yes， that is my question， thanks

**Comment by daydayfun** [comment]
> @daydayfun,
> 
> oneDNN dispatches the best available implementation for particular problem. In your particular case it seems that JIT implementation for the problem your are running is not available and oneDNN resorts to `gemm:blas` implementation. Typically this happens in "corner case" scenarios, like padding size exceeding half of the kernel size, large strides, or large kernels. If you share complete verbose line we may be able to diagnose this further.


mkldnn version: c0095ec26c
pytorch version: 1.0

I found my convolution case can dispatch to these  four functions in MKLDNN lib
mkl_gemm_convolution_fwd_t
jit_avx2_gemm_convolution_fwd_t
jit_avx512_common_gemm_convolution_fwd_t
ref_convolution_fwd_t

because of the logic :
    mkldnn::impl::primitive_desc_iterator_t &operator++() {
        if (pd_) { delete pd_; pd_ = nullptr; }
        while (++idx_ != last_idx_) {
            auto s = impl_list_[idx_](&pd_, op_desc_, &attr_, engine_,
                    hint_fwd_pd_);
            if (s ==  mkldnn::impl::status::success) break;  //？？first match   mkl_gemm_convolution_fwd_t， and results in bad convlution performace??
        }
        return *this;
    }

**Comment by daydayfun** [comment]
I think it's the problem of src/cpu/engine.cpp

#define INSTANCE(...) &primitive_desc_t::create<__VA_ARGS__::pd_t>
static const pd_create_f cpu_impl_list[] = {
    /* RNN */
    INSTANCE(ref_rnn_fwd_t),
    INSTANCE(ref_rnn_bwd_t),
    /* conv */
    INSTANCE(jit_avx512_common_dw_convolution_fwd_t),
    INSTANCE(jit_avx512_common_dw_convolution_bwd_data_t),
    INSTANCE(jit_avx512_common_1x1_convolution_fwd_f32_t),
    INSTANCE(jit_avx512_common_1x1_convolution_bwd_data_f32_t),
    INSTANCE(jit_avx512_common_1x1_convolution_bwd_weights_t),
    INSTANCE(jit_avx512_common_1x1_convolution_fwd_s16s16s32_t),
    INSTANCE(jit_avx512_common_1x1_convolution_bwd_data_s16s16s32_t),
    INSTANCE(jit_avx512_core_convolution_winograd_fwd_t),
    INSTANCE(jit_avx512_core_convolution_winograd_bwd_data_t),
    INSTANCE(jit_avx512_core_convolution_winograd_bwd_weights_t),
    INSTANCE(jit_avx512_common_convolution_winograd_fwd_t),
    INSTANCE(jit_avx512_common_convolution_winograd_bwd_data_t),
    INSTANCE(jit_avx512_common_convolution_winograd_bwd_weights_t),
    INSTANCE(jit_avx512_common_convolution_fwd_t<f32>),
    INSTANCE(jit_avx512_common_convolution_bwd_data_t<f32>),
    INSTANCE(jit_avx512_common_convolution_bwd_weights_t<f32>),
    INSTANCE(jit_avx2_dw_convolution_fwd_t),
    INSTANCE(jit_avx2_dw_convolution_bwd_data_t),
    INSTANCE(jit_avx2_1x1_convolution_fwd_t),
    INSTANCE(jit_avx2_1x1_convolution_bwd_data_t),
    INSTANCE(jit_avx2_1x1_convolution_bwd_weights_t),
    INSTANCE(jit_sse42_dw_convolution_fwd_t),
    INSTANCE(jit_sse42_dw_convolution_bwd_data_t),
    INSTANCE(jit_sse42_1x1_convolution_fwd_t),
    INSTANCE(jit_avx2_convolution_fwd_t),
    INSTANCE(jit_avx2_convolution_bwd_data_t),
    INSTANCE(jit_avx2_convolution_bwd_weights_t),
    INSTANCE(jit_sse42_convolution_fwd_t),
    INSTANCE(mkl_gemm_convolution_fwd_t),
    INSTANCE(mkl_gemm_convolution_bwd_data_t),
    INSTANCE(mkl_gemm_convolution_bwd_weights_t),
    INSTANCE(jit_avx512_common_gemm_convolution_fwd_t),
    INSTANCE(jit_avx512_common_gemm_convolution_bwd_data_t),
    INSTANCE(jit_avx512_common_gemm_convolution_bwd_weights_t),
    INSTANCE(jit_avx2_gemm_convolution_fwd_t),
    INSTANCE(jit_avx2_gemm_convolution_bwd_data_t),
    INSTANCE(jit_avx2_gemm_convolution_bwd_weights_t),
    INSTANCE(ref_convolution_fwd_t<f32>),
    INSTANCE(ref_convolution_bwd_data_t<f32, f32, f32, f32>),
    INSTANCE(ref_convolution_bwd_weights_t<f32, f32, f32, f32>),
    /* conv (int) */
    INSTANCE(jit_avx512_core_u8s8s32x_wino_convolution_fwd_t<f32>),
    INSTANCE(jit_avx512_core_u8s8s32x_wino_convolution_fwd_t<s32>),
    INSTANCE(jit_avx512_core_u8s8s32x_wino_convolution_fwd_t<s8>),
    INSTANCE(jit_avx512_core_u8s8s32x_wino_convolution_fwd_t<u8>),
    INSTANCE(jit_avx512_common_convolution_fwd_t<s16, s16, s32>),
    INSTANCE(jit_avx512_core_u8s8s32x_1x1_convolution_fwd_t<f32>),
    INSTANCE(jit_avx512_core_u8s8s32x_1x1_convolution_fwd_t<s32>),
    INSTANCE(jit_avx512_core_u8s8s32x_1x1_convolution_fwd_t<u8>),
    INSTANCE(jit_avx512_core_u8s8s32x_1x1_convolution_fwd_t<s8>),
    INSTANCE(jit_avx512_core_u8s8s32x_convolution_fwd_t<f32>),
    INSTANCE(jit_avx512_core_u8s8s32x_convolution_fwd_t<s32>),
    INSTANCE(jit_avx512_core_u8s8s32x_convolution_fwd_t<u8>),
    INSTANCE(jit_avx512_core_u8s8s32x_convolution_fwd_t<s8>),
    INSTANCE(jit_avx512_common_convolution_bwd_data_t<s16, s16, s32>),
    INSTANCE(jit_avx512_common_convolution_bwd_weights_t<s16, s16, s32>),
    INSTANCE(_gemm_u8s8s32x_convolution_fwd_t<false, s32>),
    INSTANCE(_gemm_u8s8s32x_convolution_fwd_t<false, u8>),
    INSTANCE(_gemm_u8s8s32x_convolution_fwd_t<false, s8>),
    INSTANCE(_gemm_u8s8s32x_convolution_fwd_t<false, f32>),
    INSTANCE(ref_convolution_fwd_t<s16,s16, s32, s32>),
    INSTANCE(ref_convolution_fwd_t<u8, s8, s32, s32>),
    INSTANCE(ref_convolution_fwd_t<u8, s8, s8, s32>),
    INSTANCE(ref_convolution_fwd_t<u8, s8, u8, s32>),
    INSTANCE(ref_convolution_bwd_data_t<s32, s16, s16, s32>),
    INSTANCE(ref_convolution_bwd_weights_t<s16, s32, s16, s32>),
    /* deconv */
    INSTANCE(ref_deconvolution_bwd_weights_t),
    INSTANCE(ref_deconvolution_bwd_data_t),
    INSTANCE(ref_deconvolution_fwd_t),
    /* eltwise */
    INSTANCE(jit_uni_eltwise_fwd_t<avx512_common>),
    INSTANCE(jit_uni_eltwise_bwd_t<avx512_common>),
    INSTANCE(jit_uni_eltwise_fwd_t<avx2>),
    INSTANCE(jit_uni_eltwise_bwd_t<avx2>),
    INSTANCE(jit_uni_eltwise_fwd_t<sse42>),
    INSTANCE(jit_uni_eltwise_bwd_t<sse42>),
    INSTANCE(ref_eltwise_fwd_t<f32>),
    INSTANCE(ref_eltwise_bwd_t<f32>),
    /* eltwise (int) */
    INSTANCE(ref_eltwise_fwd_t<s32>),
    INSTANCE(ref_eltwise_fwd_t<s16>),
    INSTANCE(ref_eltwise_fwd_t<s8>),
    INSTANCE(ref_eltwise_fwd_t<u8>),
    INSTANCE(ref_eltwise_bwd_t<s32>),
    INSTANCE(ref_eltwise_bwd_t<s16>),
    /* softmax */
    INSTANCE(ref_softmax_fwd_t<f32>),
    /* pool */
    INSTANCE(jit_uni_pooling_fwd_t<avx512_common>),
    INSTANCE(jit_uni_pooling_bwd_t<avx512_common>),
    INSTANCE(jit_uni_pooling_fwd_t<avx2>),
    INSTANCE(jit_uni_pooling_bwd_t<avx2>),
    INSTANCE(jit_uni_pooling_fwd_t<sse42>),
    INSTANCE(jit_uni_pooling_bwd_t<sse42>),
    INSTANCE(nchw_pooling_fwd_t<f32>),
    INSTANCE(nchw_pooling_bwd_t<f32>),
    INSTANCE(ref_pooling_fwd_t<f32>),
    INSTANCE(ref_pooling_bwd_t<f32>),
    /* pool (int) */
    INSTANCE(jit_avx512_core_i8i8_pooling_fwd_t),
    INSTANCE(ref_pooling_fwd_t<s32>),
    INSTANCE(ref_pooling_fwd_t<s16, s32>),
    INSTANCE(ref_pooling_fwd_t<s8, s32>),
    INSTANCE(ref_pooling_fwd_t<u8, s32>),
    INSTANCE(ref_pooling_bwd_t<s32>),
    INSTANCE(ref_pooling_bwd_t<s16, s32>),
    /* lrn */
    INSTANCE(jit_avx512_common_lrn_fwd_t),
    INSTANCE(jit_avx512_common_lrn_bwd_t),
    INSTANCE(jit_uni_lrn_fwd_t<avx2>),
    INSTANCE(jit_uni_lrn_bwd_t<avx2>),
    INSTANCE(jit_uni_lrn_fwd_t<sse42>),
    INSTANCE(ref_lrn_fwd_t<f32>),
    INSTANCE(ref_lrn_bwd_t<f32>),
    /* batch normalization */
    INSTANCE(jit_uni_batch_normalization_fwd_t<avx512_common>),
    INSTANCE(jit_uni_batch_normalization_bwd_t<avx512_common>),
    INSTANCE(jit_uni_batch_normalization_fwd_t<avx2>),
    INSTANCE(jit_uni_batch_normalization_bwd_t<avx2>),
    INSTANCE(jit_uni_batch_normalization_fwd_t<sse42>),
    INSTANCE(ncsp_batch_normalization_fwd_t),
    INSTANCE(ncsp_batch_normalization_bwd_t),
    INSTANCE(nspc_batch_normalization_fwd_t),
    INSTANCE(nspc_batch_normalization_bwd_t),
    INSTANCE(ref_batch_normalization_fwd_t<f32>),
    INSTANCE(ref_batch_normalization_bwd_t<f32>),
    /* inner product */
    INSTANCE(gemm_inner_product_fwd_t<f32>),
    INSTANCE(gemm_inner_product_bwd_data_t<f32>),
    INSTANCE(gemm_inner_product_bwd_weights_t<f32>),
    INSTANCE(jit_uni_inner_product_fwd_t<avx512_common>),
    INSTANCE(jit_uni_inner_product_bwd_weights_t<avx512_common>),
    INSTANCE(jit_uni_inner_product_bwd_data_t<avx512_common>),
    INSTANCE(jit_uni_inner_product_fwd_t<avx2>),
    INSTANCE(jit_uni_inner_product_bwd_weights_t<avx2>),
    INSTANCE(jit_uni_inner_product_bwd_data_t<avx2>),
    INSTANCE(ref_inner_product_fwd_t<f32>),
    INSTANCE(ref_inner_product_bwd_data_t<f32, f32, f32, f32>),
    INSTANCE(ref_inner_product_bwd_weights_t<f32>),
    /* inner product (int) */
    INSTANCE(ref_inner_product_fwd_t<s16, s16, s32, s32>),
    INSTANCE(ref_inner_product_bwd_data_t<s32, s16, s16, s32>),
    INSTANCE(ref_inner_product_fwd_t<u8, s8, u8, s32>),
    /* conv_eltwise */
    INSTANCE(jit_avx512_common_dw_convolution_relu_t),
    INSTANCE(jit_avx512_common_convolution_winograd_relu_t),
    INSTANCE(jit_avx512_common_1x1_convolution_relu_f32_t),
    INSTANCE(jit_avx512_common_convolution_relu_t<f32>),
    INSTANCE(jit_avx2_dw_convolution_relu_t),
    INSTANCE(jit_avx2_1x1_convolution_relu_t),
    INSTANCE(jit_sse42_dw_convolution_relu_t),
    INSTANCE(jit_sse42_1x1_convolution_relu_t),
    INSTANCE(jit_avx2_convolution_relu_t),
    INSTANCE(jit_sse42_convolution_relu_t),
    INSTANCE(mkl_gemm_convolution_relu_t),
    INSTANCE(jit_avx512_common_gemm_convolution_relu_t),
    INSTANCE(jit_avx2_gemm_convolution_relu_t),
    INSTANCE(ref_convolution_relu_t<f32>),
    /* conv_eltwise (int) */
    INSTANCE(jit_avx512_core_u8s8s32x_wino_convolution_relu_t<f32>),
    INSTANCE(jit_avx512_core_u8s8s32x_wino_convolution_relu_t<s32>),
    INSTANCE(jit_avx512_core_u8s8s32x_wino_convolution_relu_t<s8>),
    INSTANCE(jit_avx512_core_u8s8s32x_wino_convolution_relu_t<u8>),
    INSTANCE(jit_avx512_common_1x1_convolution_relu_s16s16s32_t),
    INSTANCE(jit_avx512_common_convolution_relu_t<s16, s16, s32>),
    INSTANCE(jit_avx512_core_u8s8s32x_1x1_convolution_relu_t<f32>),
    INSTANCE(jit_avx512_core_u8s8s32x_1x1_convolution_relu_t<s32>),
    INSTANCE(jit_avx512_core_u8s8s32x_1x1_convolution_relu_t<s8>),
    INSTANCE(jit_avx512_core_u8s8s32x_1x1_convolution_relu_t<u8>),
    INSTANCE(jit_avx512_core_u8s8s32x_convolution_relu_t<f32>),
    INSTANCE(jit_avx512_core_u8s8s32x_convolution_relu_t<s32>),
    INSTANCE(jit_avx512_core_u8s8s32x_convolution_relu_t<u8>),
    INSTANCE(jit_avx512_core_u8s8s32x_convolution_relu_t<s8>),
    INSTANCE(_gemm_u8s8s32x_convolution_fwd_t<true, s32>),
    INSTANCE(_gemm_u8s8s32x_convolution_fwd_t<true, u8>),
    INSTANCE(_gemm_u8s8s32x_convolution_fwd_t<true, s8>),
    INSTANCE(_gemm_u8s8s32x_convolution_fwd_t<true, f32>),
    INSTANCE(ref_convolution_relu_t<s16, s16, s32, s32>),
    INSTANCE(ref_convolution_relu_t<u8, s8, s32, s32>),
    INSTANCE(ref_convolution_relu_t<u8, s8, s8, s32>),
    INSTANCE(ref_convolution_relu_t<u8, s8, u8, s32>),
    /* eol */
    nullptr,
};
#undef INSTANCE
}

**Comment by daydayfun** [comment]
I change the position  of  these two func pointers, and my conv runs in  gemm:avx2 alogrithm
INSTANCE(mkl_gemm_convolution_fwd_t), 
INSTANCE(jit_avx2_gemm_convolution_fwd_t),

**Comment by rsdubtso** [comment]
> mkldnn version: c0095ec
pytorch version: 1.0

This mkldnn is very old. There were tons of updates to GEMM since then.

The distinction between `mkl_gemm_convolution_fwd_t` and `jit_avx2_gemm_convolution_fwd_t` was that one called into MKL and the other used mkldnn GEMM implementation. The MKL version also dispatched to AVX2 on capable hardware, but it was transparent to the mkldnn library so it was not noted in the implementation name.

The latest mkldnn (now called oneDNN) relies on its own GEMM implementation only, so the MKL version went away. oneDNN GEMM also dispatches to ISA-specifica implementation based on the hardware capabilities.

**Comment by daydayfun** [comment]
Thanks very much, pytorch 1.0 uses the very old mkldnn, and I found pytorch 1.0.1 uses the newer mkldnn and the cpu_impl_list has been changed.
Anyway, my question solved , thanks all.
The mkldnn source code is very hard to debug. Can you give me some doc. I am interested in the source code
